Source/Core/DiscIO/FileSystemGCWii.cpp

```diff
@@ -19,307 +21,444 @@
 	}
 
-	// Iterate through the 11 data segments
-	for (u8 i = 0; i < 11; i++)
+	// Just in case the last string in the name table isn't null terminated
+	m_file_system_table[fst_size - 1] = 0;
+
+	// Create the root object
+	m_root = CFileInfoGCWii(m_file_system_table.data(), m_wii, raw_fst_size);
+
+	if (!m_root.IsDirectory())
 	{
-		offset = m_rVolume->Read32(DolOffset + 0x1c + i * 4, m_Wii);
-		size   = m_rVolume->Read32(DolOffset + 0xac + i * 4, m_Wii);
-		if (offset + size > DolSize)
-			DolSize = offset + size;
+		ERROR_LOG(DISCIO, "File system root is not a directory");
+		return;
 	}
-	return DolSize;
+
+	if (m_root.GetSize() * 0xC > fst_size)
+	{
+		ERROR_LOG(DISCIO, "File system has too many entries for its size");
+		return;
+	}
+
+	// If we haven't returned yet, everything succeeded
+	m_valid = true;
 }
 
-bool CFileSystemGCWii::GetBootDOL(u8* &buffer, u32 DolSize) const
+CFileSystemGCWii::~CFileSystemGCWii()
 {
-	u32 DolOffset = m_rVolume->Read32(0x420, m_Wii) << GetOffsetShift();
-	return m_rVolume->Read(DolOffset, DolSize, buffer, m_Wii);
 }
 
-bool CFileSystemGCWii::ExportDOL(const std::string& _rExportFolder) const
+const IFileInfo& CFileSystemGCWii::GetRoot() const
 {
-	u32 DolOffset = m_rVolume->Read32(0x420, m_Wii) << GetOffsetShift();
-	u32 DolSize = GetBootDOLSize();
+	return m_root;
+}
 
-	std::vector<u8> buffer(DolSize);
-	if (m_rVolume->Read(DolOffset, DolSize, &buffer[0], m_Wii))
+std::unique_ptr<IFileInfo> CFileSystemGCWii::FindFileInfo(const std::string& path) const
+{
+	if (!IsValid())
+		return nullptr;
+
+	return FindFileInfo(path, m_root);
+}
+
+std::unique_ptr<IFileInfo> CFileSystemGCWii::FindFileInfo(const std::string& path, const IFileInfo& file_info) const
+{
+	// Given a path like "directory1/directory2/fileA.bin", this method will
+	// find directory1 and then call itself to search for "directory2/fileA.bin".
+
+	// If there's nothing left to search for, return the current entry
+	if (path.empty())
+		return std::unique_ptr<IFileInfo>(new CFileInfoGCWii((const CFileInfoGCWii&)file_info));
+
+	// It's only possible to search in directories. Searching in a file is an error
+	if (!file_info.IsDirectory())
+		return nullptr;
+
+	std::string searching_for;
+	std::string rest_of_path;
+	size_t first_dir_separator = path.find('/');
+	if (first_dir_separator != std::string::npos)
 	{
-		std::string exportName(_rExportFolder + "/boot.dol");
+		searching_for = path.substr(0, first_dir_separator);
+		rest_of_path = path.substr(first_dir_separator + 1);
+	}
+	else
+	{
+		searching_for = path;
+		rest_of_path = "";
+	}
 
-		File::IOFile DolFile(exportName, "wb");
-		if (DolFile)
+	for (const IFileInfo& child : file_info)
+	{
+		if (child.GetName() == searching_for)
 		{
-			DolFile.WriteBytes(&buffer[0], DolSize);
-			return true;
+			// A match is found. The rest of the path is passed on to finish the search
+			std::unique_ptr<IFileInfo> result = FindFileInfo(rest_of_path, child);
+			if (result)
+				return result;
```

Comment:
couldn't you just do 

```cpp
return FindFileInfo(rest_of_path, child);
```

considering the fail case below this returns nullptr anyway?